node/src/blockchain/blockchain_interface/blockchain_interface_web3/lower_level_interface_web3.rs

```diff
@@ -1,395 +1,762 @@
+        address: Address,
+    ) -> Box<dyn Future<Item = U256, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .balance(address, None)
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
     }
 
-    fn get_service_fee_balance(&self, wallet: &Wallet) -> ResultForBalance {
-        self.contract
-            .query(
-                "balanceOf",
-                wallet.address(),
-                None,
-                Options::default(),
-                None,
-            )
-            .map_err(|e| BlockchainError::QueryFailed(format!("{} for wallet {}", e, wallet)))
-            .wait()
+    fn get_service_fee_balance(
+        &self,
+        address: Address,
+    ) -> Box<dyn Future<Item = U256, Error = BlockchainError>> {
+        Box::new(
+            self.contract
+                .query("balanceOf", address, None, Options::default(), None)
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
     }
 
-    fn get_block_number(&self) -> LatestBlockNumber {
-        self.web3
-            .eth()
-            .block_number()
-            .map_err(|e| BlockchainError::QueryFailed(e.to_string()))
-            .wait()
+    fn get_gas_price(&self) -> Box<dyn Future<Item = U256, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .gas_price()
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
     }
 
-    fn get_transaction_id(&self, wallet: &Wallet) -> ResultForNonce {
-        self.web3
-            .eth()
-            .transaction_count(wallet.address(), Some(BlockNumber::Pending))
-            .map_err(|e| BlockchainError::QueryFailed(format!("{} for wallet {}", e, wallet)))
-            .wait()
+    fn get_block_number(&self) -> Box<dyn Future<Item = U64, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .block_number()
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
+    }
+
+    fn get_transaction_id(
+        &self,
+        address: Address,
+    ) -> Box<dyn Future<Item = U256, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .transaction_count(address, Some(BlockNumber::Pending))
+                .map_err(move |e| QueryFailed(format!("{} for wallet {}", e, address))),
+        )
+    }
+
+    fn get_transaction_receipt(
+        &self,
+        hash: H256,
+    ) -> Box<dyn Future<Item = Option<TransactionReceipt>, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .transaction_receipt(hash)
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
+    }
+
+    fn get_transaction_receipt_batch(
+        &self,
+        hash_vec: Vec<H256>,
+    ) -> Box<dyn Future<Item = Vec<TransactionReceiptResult>, Error = BlockchainError>> {
```

Comment:
The returned native web3 structure is quite big. Keeping it just to read one tiny field later on might not be what we want. But I don't think I have the right feeling for it at the moment. 

**Maybe Dan could help us out with a sword to this Gordian knot...**  

Other arguments:

a) You guys even had to make an extra builder among the test utils just to construct this huge, complicated structure of which we need less then 10% information in it. The builder itself takes up quite a lot of space, which could've been saved from this, in the purest honesty.

b) And more importantly, you must not use any third-party data structure in your response because that would mean it couldn't be transposed to a different chain in its implementation. You will need to create your own `struct` and when you're dirty already, I really don't see a reason why it should hold so many information. It basically should only answer if to the questions we are interested in. No additional information which would be without a use. If such a need arises I'm sure it would be in a distant future but not now and we will ever be able to extend the list of parameters we would rely on from the transaction.

Reply:
I made sure this is referred properly from another comment.